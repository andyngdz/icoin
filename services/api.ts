import { Format } from 'services'
import { IAPI, IAssetResponse } from 'types'
import { PER_PAGE } from 'data'
import axios from 'axios'

const CoingeckoAPI = axios.create({
  baseURL: process.env.NEXT_PUBLIC_COINGECKO_API
})

const CoinCapAPI = axios.create({
  baseURL: process.env.NEXT_PUBLIC_COINCAP_API
})

const API: IAPI = {
  getGlobals: () => CoingeckoAPI.get('/global'),

  getAssets: async (
    params = {
      limit: PER_PAGE
    }
  ) => {
    const response = await CoinCapAPI.get<IAssetResponse>('/assets', {
      params
    })
    const { data } = response.data

    response.data.data = Format.toNumber(data, [
      'marketCapUsd',
      'priceUsd',
      'volumeUsd24Hr',
      'changePercent24Hr',
      'supply',
      'maxSupply',
      'vwap24Hr'
    ])

    return response
  }
}

export { API }
