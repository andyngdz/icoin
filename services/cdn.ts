const CDN = {
  getIcon: (id: string) =>
    `https://cdn.jsdelivr.net/gh/atomiclabs/cryptocurrency-icons@9ab8d6934b83a4aa8ae5e8711609a70ca0ab1b2b/svg/color/${id}.svg`,

  defaultIcon: `https://cdn.jsdelivr.net/gh/atomiclabs/cryptocurrency-icons@9ab8d6934b83a4aa8ae5e8711609a70ca0ab1b2b/svg/color/generic.svg`,
}

export { CDN }
