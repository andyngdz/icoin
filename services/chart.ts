/**
 * Author: Andy D. Ng <itc.anhduy@gmail.com>
 * Chart.js contains the useful function to manage a Chart
 */

import { ChartData, ChartOptions, TimeScale } from 'chart.js'
import { IAssetHistory, ICalculateInterval, TTime } from 'types'
import { merge } from 'lodash'
import { sub } from 'date-fns'
import { TIME_TO_INTERVAL } from 'data'
import ChartJS from 'chart.js'

/**
 * @description
 * We are using Roboto font and
 * Set it here for the ChartJS as well
 */
ChartJS.defaults.global.defaultFontFamily = "'Inter', sans-serif"

const Chart = {
  /**
   * @description Calculate interval post data by id, time, start and end date
   *
   * @param id The id of coin such as: bitcoin, ethereum
   *
   * @param time The time from the TimeSelection component - (Check TTime in chart.d.ts)
   *
   * @returns `ICalculateInterval`
   *
   * @example
   * ```
   *    const { time, onTimeChange } = useTime()
   *     const [getIntervals, { data }] = useLazyQuery<IAssetHistories>(COIN_CHART)
   *
   *     useEffect(() => {
   *       const interval = Chart.calculateInterval(id, time)
   *       getIntervals({ variables: interval })
   *     }, [time])
   *```
   */
  calculateInterval: (id: string, time: TTime): ICalculateInterval => {
    let subtractDate: Date
    const end = Date.now()

    switch (time) {
      case '1D':
        subtractDate = sub(end, { days: 1 })
        break
      case '1W':
        subtractDate = sub(end, { weeks: 1 })
        break
      case '1M':
        subtractDate = sub(end, { months: 1 })
        break
      case '3M':
        subtractDate = sub(end, { months: 3 })
        break
      case '6M':
        subtractDate = sub(end, { months: 6 })
        break
      case '1Y':
      case 'ALL':
        subtractDate = sub(end, { years: 1 })
        break
    }

    return {
      end,
      id,
      interval: TIME_TO_INTERVAL[time],
      start: subtractDate.valueOf()
    }
  },

  /**
   * @description Create time scale with unit for Line Chart
   *
   * @param time The time from the TimeSelection component - (Check TTime in chart.d.ts)
   *
   * @returns `TimeScale`
   */
  createTimeScale: (time: TTime): TimeScale => {
    const timeScale: TimeScale = {}

    switch (time) {
      case '1D':
        timeScale.unit = 'hour'
        break
      case '1W':
        timeScale.unit = 'day'
        break
      case '1M':
      case '3M':
        timeScale.unit = 'week'
        break
      case '6M':
      case '1Y':
      case 'ALL':
        timeScale.unit = 'month'
        break
    }

    return timeScale
  },

  /**
   * @description Create a new chart
   *
   * @param ctx The canvas rendering context
   *
   * @param assetHistories Contains the list history data such as: timestamp, priceUsd, etc...
   *
   * @param options The chart options, it will be merged with the default options
   *
   * @returns `ChartJS`
   */
  createNewChart: (
    canvas: HTMLCanvasElement,
    time: TTime,
    chartData: ChartData,
    options?: ChartOptions
  ): ChartJS => {
    const ctx = canvas.getContext('2d')

    return new ChartJS(ctx, {
      type: 'line',
      data: chartData,
      options: merge(
        {
          tooltips: {
            position: 'nearest'
          },
          legend: {
            display: false
          },
          scales: {
            xAxes: [
              {
                type: 'time',
                time: Chart.createTimeScale(time),
                gridLines: {
                  display: false
                }
              }
            ],
            yAxes: [
              {
                position: 'right',
                gridLines: {
                  drawBorder: false
                }
              }
            ]
          }
        },
        options
      )
    })
  },

  /**
   * @description Create chart data
   *
   * @param assetHistories Contains the list history data such as: timestamp, priceUsd, etc...
   *
   * @returns `ChartData`
   */
  createChartData: (assetHistories: Array<IAssetHistory>): ChartData => {
    const data = assetHistories.map(({ timestamp, priceUsd }) => ({
      t: timestamp,
      y: priceUsd
    }))

    const chartData: ChartData = {
      datasets: [
        {
          data
        }
      ]
    }

    return chartData
  }
}

export { Chart }
