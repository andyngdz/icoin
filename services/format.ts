import { formatCurrency } from '@coingecko/cryptoformat'
import numeral from 'numeral'

const numberFormatter = new Intl.NumberFormat('en-US')

const percentFormatter = new Intl.NumberFormat('en-US', {
  style: 'percent',
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
})

const Format = {
  bigNumber: (value: number) => numeral(value).format('($0.00a)'),

  currency: (value: number) => formatCurrency(value, 'USD', 'en'),

  percent: (value: number) => percentFormatter.format(value / 100),
}

export { Format, numberFormatter, percentFormatter }
