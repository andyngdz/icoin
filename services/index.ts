export * from './paginate'
export * from './chart'
export * from './format'
export * from './api'
export * from './cdn'
export * from './test-utils'
export * from './price-direction'
export * from './price'
export * from './routes'
