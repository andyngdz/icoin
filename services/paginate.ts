const Paginate = {
  offset: (page: number, perPage: number): number => --page * perPage,

  count: (total: number, perPage: number): number => Math.ceil(total / perPage)
}

export { Paginate }
