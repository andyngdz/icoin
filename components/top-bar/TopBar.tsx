import { API, Format } from 'services'
import { makeStyles, Typography, AppBar, Toolbar } from '@material-ui/core'
import { useAsync } from 'react-use'

const useStyles = makeStyles(
  (theme) => ({
    topbar: {
      display: 'flex',
      justifyContent: 'space-between',
    },

    info: {
      '& > span': {
        paddingRight: theme.spacing(2),
        fontWeight: theme.typography.button.fontWeight,
      },
    },
  }),
  {
    name: 'TopBar',
  }
)

const TopBar = () => {
  const classes = useStyles()
  const { loading, value } = useAsync(API.getGlobals)

  if (!loading) {
    const { data } = value.data

    return (
      <AppBar color="default" variant="outlined">
        <Toolbar className={classes.topbar}>
          <div className={classes.info}>
            <Typography variant="subtitle2" component="span">
              Cryptocurrencies:&nbsp;
              <Typography variant="subtitle2" component="span" color="primary">
                {data.active_cryptocurrencies}
              </Typography>
            </Typography>
            <Typography variant="subtitle2" component="span">
              Markets:&nbsp;
              <Typography variant="subtitle2" component="span" color="primary">
                {data.markets}
              </Typography>
            </Typography>
          </div>
          <div className={classes.info}>
            <Typography variant="subtitle2" component="span">
              Market Cap:&nbsp;
              <Typography variant="subtitle2" component="span" color="primary">
                {Format.currency(data.total_market_cap.usd)}
              </Typography>
            </Typography>
            <Typography variant="subtitle2" component="span">
              24h Volume:&nbsp;
              <Typography variant="subtitle2" component="span" color="primary">
                {Format.currency(data.total_volume.usd)}
              </Typography>
            </Typography>
            <Typography variant="subtitle2" component="span">
              Dominance:&nbsp;
              <Typography variant="subtitle2" component="span" color="primary">
                BTC {data.market_cap_percentage.btc.toFixed(2)}%
              </Typography>
            </Typography>
            <Typography variant="subtitle2" component="span" color="primary">
              ETH {data.market_cap_percentage.eth.toFixed(2)}%
            </Typography>
          </div>
        </Toolbar>
      </AppBar>
    )
  }

  return <></>
}

export { TopBar }
