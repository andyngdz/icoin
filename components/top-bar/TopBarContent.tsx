import { Format } from 'services'
import { IGlobalData } from 'types'
import { makeStyles, Typography } from '@material-ui/core'

interface ITopBarContentProps {
  globalData: IGlobalData
}

const useStyles = makeStyles(
  theme => ({
    content: {
      display: 'flex',
      flex: 1
    },

    info: {
      '& > span': {
        paddingRight: theme.spacing(2)
      }
    }
  }),
  {
    name: 'TopBarContent'
  }
)

const TopBarContent: React.FC<ITopBarContentProps> = ({ globalData }) => {
  const classes = useStyles()
  const { marketTotal, asset } = globalData

  if (marketTotal && asset) {
    const { assets, markets, exchangeVolumeUsd24Hr } = marketTotal
    const marketCap = Format.toNumber(marketTotal, 'marketCapUsd')
    const bitcoinCap = Format.toNumber(asset, 'marketCapUsd')

    return (
      <div className={classes.content}>
        <div className={classes.info}>
          <Typography variant="overline" component="span">
            Assets:&nbsp;
            <Typography variant="overline" component="span" color="primary">
              {assets}
            </Typography>
          </Typography>
          <Typography variant="overline" component="span">
            Markets:&nbsp;
            <Typography variant="overline" component="span" color="primary">
              {markets}
            </Typography>
          </Typography>
          <Typography variant="overline" component="span">
            Market Cap:&nbsp;
            <Typography variant="overline" component="span" color="primary">
              {Format.currency(marketCap)}
            </Typography>
          </Typography>
          <Typography variant="overline" component="span">
            24h Volume:&nbsp;
            <Typography variant="overline" component="span" color="primary">
              {Format.currency(exchangeVolumeUsd24Hr)}
            </Typography>
          </Typography>
          <Typography variant="overline" component="span">
            Dominance:&nbsp;
            <Typography variant="overline" component="span" color="primary">
              BTC {Format.percent((bitcoinCap * 100) / marketCap)}
            </Typography>
          </Typography>
        </div>
      </div>
    )
  }

  return <></>
}

export { TopBarContent }
