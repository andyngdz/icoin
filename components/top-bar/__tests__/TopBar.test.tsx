import { API } from 'services'
import {
  act,
  customRender,
  mockResponse,
  RenderResult,
} from 'services/test-utils'
import { TopBar } from 'components/top-bar/TopBar'
import Globals from 'mock-data/globals.json'

describe('TopBar', () => {
  it('Should render correctly', async () => {
    let renderResult: RenderResult

    await act(async () => {
      renderResult = customRender(<TopBar />)
    })

    expect(API.getGlobals).toHaveBeenCalled()
    expect(API.getGlobals).toHaveReturnedWith(mockResponse(Globals))
    expect(renderResult).toMatchSnapshot()
  })
})
