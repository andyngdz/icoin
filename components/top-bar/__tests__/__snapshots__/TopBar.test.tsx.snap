// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar Should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorDefault mui-fixed MuiPaper-outlined"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular TopBar-topbar-1 MuiToolbar-gutters"
        >
          <div
            class="TopBar-info-2"
          >
            <span
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Cryptocurrencies: 
              <span
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
              >
                6191
              </span>
            </span>
            <span
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Markets: 
              <span
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
              >
                564
              </span>
            </span>
          </div>
          <div
            class="TopBar-info-2"
          >
            <span
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Market Cap: 
              <span
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
              >
                $1,016,628,962,622.60
              </span>
            </span>
            <span
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              24h Volume: 
              <span
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
              >
                $162,857,223,748.66
              </span>
            </span>
            <span
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              Dominance: 
              <span
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
              >
                BTC 
                62.02
                %
              </span>
            </span>
            <span
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
            >
              ETH 
              15.17
              %
            </span>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorDefault mui-fixed MuiPaper-outlined"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular TopBar-topbar-1 MuiToolbar-gutters"
      >
        <div
          class="TopBar-info-2"
        >
          <span
            class="MuiTypography-root MuiTypography-subtitle2"
          >
            Cryptocurrencies: 
            <span
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
            >
              6191
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-subtitle2"
          >
            Markets: 
            <span
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
            >
              564
            </span>
          </span>
        </div>
        <div
          class="TopBar-info-2"
        >
          <span
            class="MuiTypography-root MuiTypography-subtitle2"
          >
            Market Cap: 
            <span
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
            >
              $1,016,628,962,622.60
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-subtitle2"
          >
            24h Volume: 
            <span
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
            >
              $162,857,223,748.66
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-subtitle2"
          >
            Dominance: 
            <span
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
            >
              BTC 
              62.02
              %
            </span>
          </span>
          <span
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorPrimary"
          >
            ETH 
            15.17
            %
          </span>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
